substrate/frame/beefy/src/equivocation.rs

```diff
@@ -106,10 +108,14 @@ where
 		self.time_slot
 	}
 
-	// The formula is min((3k / n)^2, 1)
-	// where k = offenders_number and n = validators_number
 	fn slash_fraction(&self, offenders_count: u32) -> Perbill {
-		// Perbill type domain is [0, 1] by definition
+		if let Some(slash_fraction) = self.hardcoded_slash_fraction {
+			return slash_fraction;
+		}
+
+		// `Perbill` type domain is [0, 1] by definition
+		// The formula is min((3k / n)^2, 1)
+		// where k = offenders_number and n = validators_number
 		Perbill::from_rational(3 * offenders_count, self.validator_set_count).square()
```

Comment:
@serban300 just out of curiosity, where is `min( , 1)` applied?

Reply:
I think the `Perbill` implementation takes care of this in all the constructors and operations. For example:
https://github.com/paritytech/polkadot-sdk/blob/f2835be96a89f6ba684b657be931fc52a946d0ad/substrate/primitives/arithmetic/src/per_things.rs#L378
https://github.com/paritytech/polkadot-sdk/blob/f2835be96a89f6ba684b657be931fc52a946d0ad/substrate/primitives/arithmetic/src/per_things.rs#L704
https://github.com/paritytech/polkadot-sdk/blob/f2835be96a89f6ba684b657be931fc52a946d0ad/substrate/primitives/arithmetic/src/per_things.rs#L917-L933